### PR TITLE
Fail build when unused c code is detected

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

To keep our code clean we should fail the build when unused c code is detected.

Modifications:

- Add '-Wunused-variable' to build flags
- Removed some unused code, added ifdefs

Result:

Cleaner code.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@nmittler @Scottmitch @davidben PTAL

File: openssl-dynamic/src/main/c/sslcontext.c [hunk 1]
```diff
@@ -1350,14 +1353,10 @@ static int cert_requested(SSL* ssl, X509** x509Out, EVP_PKEY** pkeyOut) {
...
     jlong certChain;
-    jlong privateKey;
     tcn_get_java_env(&e);
     int certChainLen;
     int i;
-
-#if defined(LIBRESSL_VERSION_NUMBER)
-    return -1;
-#else
+    jlong privateKey;
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/sslcontext.c]: can this move back with the other `jlong`?

File: openssl-dynamic/src/main/c/sslcontext.c [hunk 2]
```diff
@@ -1350,14 +1353,10 @@ static int cert_requested(SSL* ssl, X509** x509Out, EVP_PKEY** pkeyOut) {
     X509 *cert = NULL;
     EVP_PKEY* pkey = NULL;
     jlong certChain;
-    jlong privateKey;
     tcn_get_java_env(&e);
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/sslcontext.c]: move this after `JNIEnv *e;` or below the variable declarations.

File: openssl-dynamic/src/main/c/ssl_private.h [hunk 3]
```diff
@@ -125,7 +125,7 @@
 
 #define MAX_ALPN_NPN_PROTO_SIZE 65535
 
-static const char* UNKNOWN_AUTH_METHOD = "UNKNOWN";
+#define UNKNOWN_AUTH_METHOD "UNKNOWN"
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/ssl_private.h]: we shouldn't have to loose the type safety. Lets do this instead (note I added the `TCN_` prefix to provide more scope and reduce the chance of collisions):

```diff
diff --git a/openssl-dynamic/src/main/c/ssl_private.h b/openssl-dynamic/src/main/c/ssl_private.h
index 1c9e4c4..f4b1338 100644
--- a/openssl-dynamic/src/main/c/ssl_private.h
+++ b/openssl-dynamic/src/main/c/ssl_private.h
@@ -125,7 +125,7 @@
 
 #define MAX_ALPN_NPN_PROTO_SIZE 65535
 
-#define UNKNOWN_AUTH_METHOD "UNKNOWN"
+extern const char* TCN_UNKNOWN_AUTH_METHOD;
 
 /* ECC: make sure we have at least 1.0.0 */
 #if !defined(OPENSSL_NO_EC) && defined(TLSEXT_ECPOINTFORMAT_uncompressed)
diff --git a/openssl-dynamic/src/main/c/sslcontext.c b/openssl-dynamic/src/main/c/sslcontext.c
index 2fa688d..bdcf38e 100644
--- a/openssl-dynamic/src/main/c/sslcontext.c
+++ b/openssl-dynamic/src/main/c/sslcontext.c
@@ -1183,7 +1183,7 @@ static const char* authentication_method(const SSL* ssl) {
             ciphers = SSL_get_ciphers(ssl);
             if (ciphers == NULL || sk_SSL_CIPHER_num(ciphers) <= 0) {
                 // No cipher available so return UNKNOWN.
-                return UNKNOWN_AUTH_METHOD;
+                return TCN_UNKNOWN_AUTH_METHOD;
             }
             return SSL_cipher_authentication_method(sk_SSL_CIPHER_value(ciphers, 0));
         }
diff --git a/openssl-dynamic/src/main/c/sslutils.c b/openssl-dynamic/src/main/c/sslutils.c
index 77cbf94..e5eb501 100644
--- a/openssl-dynamic/src/main/c/sslutils.c
+++ b/openssl-dynamic/src/main/c/sslutils.c
@@ -69,6 +69,8 @@
 #define SSL_aGOST94     0x00000100L /* GOST R 34.10-94 signature auth */
 #define SSL_aGOST01     0x00000200L /* GOST R 34.10-2001 signature auth */
 
+const char* TCN_UNKNOWN_AUTH_METHOD = "UNKNOWN";
+
 /* OpenSSL end */
 
 /*
@@ -96,7 +98,7 @@ const char* SSL_cipher_authentication_method(const SSL_CIPHER* cipher){
         case SSL_aNULL:
             return SSL_TXT_DH "_anon";
         default:
-            return UNKNOWN_AUTH_METHOD;
+            return TCN_UNKNOWN_AUTH_METHOD;
             }
     case SSL_kKRB5:
         return SSL_TXT_KRB5;
@@ -114,10 +116,10 @@ const char* SSL_cipher_authentication_method(const SSL_CIPHER* cipher){
         case SSL_aNULL:
             return SSL_TXT_ECDH "_anon";
         default:
-            return UNKNOWN_AUTH_METHOD;
+            return TCN_UNKNOWN_AUTH_METHOD;
             }
     default:
-        return UNKNOWN_AUTH_METHOD;
+        return TCN_UNKNOWN_AUTH_METHOD;
     }
 #else
     return SSL_CIPHER_get_kx_name(cipher);
```

File: openssl-dynamic/src/main/native-package/configure.ac [hunk 4]
```diff
@@ -28,7 +28,7 @@ AC_CONFIG_HEADERS([src/config.h])
 AC_CANONICAL_HOST
 AC_CANONICAL_SYSTEM
 
-${CFLAGS="-O3 -Werror -fno-omit-frame-pointer"}
+${CFLAGS="-O3 -Werror -fno-omit-frame-pointer -Wunused-variable"}
```
Comment (davidben) [inline openssl-dynamic/src/main/native-package/configure.ac]: If you feel extra adventurous later on, `-Wall` is fun. (`-Wall` implies `-Wunused-variable` so you won't need both.)

**Comment by normanmaurer** [comment]
@Scottmitch PTAL again... addressed your comments

**Comment by normanmaurer** [comment]
Cherry-picked into master (96a9665189689df7615362390e3b02ea22ef1707)

**Pull request closed without merging**